File: petties_mobile/lib/data/services/staff_shift_service.dart
import '../models/staff_shift.dart';
import 'api_client.dart';

class StaffShiftService {
  final ApiClient _apiClient;

  StaffShiftService({ApiClient? apiClient})
      : _apiClient = apiClient ?? ApiClient();

  /// Get shifts for the currently logged-in staff
  Future<List<StaffShiftResponse>> getMyShifts({
    required String startDate,
    required String endDate,
  }) async {
    final response = await _apiClient.get(
      '/shifts/me',
      queryParameters: {
        'startDate': startDate,
        'endDate': endDate,
      },
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = response.data;
      return data.map((json) => StaffShiftResponse.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load shifts');
    }
  }

  /// Get shift detail with slots
  Future<StaffShiftResponse> getShiftDetail(String shiftId) async {
    final response = await _apiClient.get('/shifts/$shiftId');

    if (response.statusCode == 200) {
      return StaffShiftResponse.fromJson(response.data);
    } else {
      throw Exception('Failed to load shift detail');
    }
  }

  /// Block a slot
  Future<SlotResponse> blockSlot(String slotId) async {
    final response = await _apiClient.patch('/slots/$slotId/block');

    if (response.statusCode == 200) {
      return SlotResponse.fromJson(response.data);
    } else {
      throw Exception('Failed to block slot');
    }
  }

  /// Unblock a slot
  Future<SlotResponse> unblockSlot(String slotId) async {
    final response = await _apiClient.patch('/slots/$slotId/unblock');

    if (response.statusCode == 200) {
      return SlotResponse.fromJson(response.data);
    } else {
      throw Exception('Failed to unblock slot');
    }
  }
}

File: petties-web/src/services/api/staffShiftService.ts
import { apiClient } from './client'
import type { StaffShiftRequest, StaffShiftResponse, SlotResponse } from '../../types/staffshift'

export const staffShiftService = {
    /**
     * Create one or more staff shifts
     */
    createShift: async (clinicId: string, data: StaffShiftRequest): Promise<StaffShiftResponse[]> => {
        const response = await apiClient.post<StaffShiftResponse[]>(`/clinics/${clinicId}/shifts`, data)
        return response.data
    },

    /**
     * Get all shifts for a clinic in a date range
     */
    getShiftsByClinic: async (
        clinicId: string,
        startDate: string,
        endDate: string
    ): Promise<StaffShiftResponse[]> => {
        const params = new URLSearchParams({
            startDate,
            endDate,
        })
        const url = `/clinics/${clinicId}/shifts?${params.toString()}`
        const response = await apiClient.get<StaffShiftResponse[]>(url)
        return response.data
    },

    /**
     * Get shifts for the currently logged-in staff
     */
    getMyShifts: async (startDate: string, endDate: string): Promise<StaffShiftResponse[]> => {
        const params = new URLSearchParams({ startDate, endDate })
        const response = await apiClient.get<StaffShiftResponse[]>(`/shifts/me?${params.toString()}`)
        return response.data
    },

    /**
     * Get shift detail with slots
     */
    getShiftDetail: async (shiftId: string): Promise<StaffShiftResponse> => {
        const response = await apiClient.get<StaffShiftResponse>(`/shifts/${shiftId}`)
        return response.data
    },

    /**
     * Delete a shift
     */
    deleteShift: async (shiftId: string): Promise<void> => {
        await apiClient.delete(`/shifts/${shiftId}`)
    },

    /**
     * Delete multiple shifts
     */
    bulkDeleteShifts: async (shiftIds: string[]): Promise<void> => {
        await apiClient.delete('/shifts/bulk', { data: shiftIds })
    },

    /**
     * Block a slot
     */
    blockSlot: async (slotId: string): Promise<SlotResponse> => {
        const response = await apiClient.patch<SlotResponse>(`/slots/${slotId}/block`)
        return response.data
    },

    /**
     * Unblock a slot
     */
    unblockSlot: async (slotId: string): Promise<SlotResponse> => {
        const response = await apiClient.patch<SlotResponse>(`/slots/${slotId}/unblock`)
        return response.data
    },
}
